### PR TITLE
Create polymers

### LINKED ISSUE #2687: Re-implement polymer creation

### PR DESCRIPTION
**Pull request by pkreissl**
Fixes #2687 

Description of changes:
 - Split diamond creation from polymer creation, put into its own file.
 - Re-implement `create_polymer`, now as ~`polymer_positions()`~ `polymer.position()` passes the
   determined positions to the interface where they are returned as numpy array.
   The actual creation now happens in the pypresso scripts.
 - Update documentation, tutorials, test cases.

PR Checklist
------------
 - [x] Tests?
   - [x] Interface
   - [x] Core 
 - [x] Docs?

### CODE REVIEW

File: doc/sphinx/particles.rst [hunk 1]
```diff
@@ -196,39 +196,60 @@ Setting up polymer chains
...
-A function that allows to create a number of polymers and polyelectrolytes.
-See :attr:`espressomd.polymer.create_polymer()` for a detailed list of
-arguments.
-
-The distance between adjacent monomers
+    from espressomd.polymer import polymer_positions
+
+If you want to have polymers in your system, this function can provide
+you with suitable particle positions.
+See :attr:`espressomd.polymer.polymer_positions()` for a detailed list of
```
Comment (KaiSzuttor) [inline doc/sphinx/particles.rst]: list of ...

File: doc/sphinx/particles.rst [hunk 2]
```diff
@@ -196,39 +196,60 @@ Setting up polymer chains
...
+Determining suitable particle positions pseudo-randomly requires the use of
+a pseudo-random number genererator, which has to be seeded. This ``seed``
+is therefore also a mandatory parameter.
+
+The function :attr:`espressomd.polymer.polymer_positions()` returns a
+three-dimensional numpy array, namely a list of polymers containing the
+positions of monomers (x, y, z). A quick example of how to set up polymers::
+
+     import espressomd
+     from espressomd.polymer import polymer_positions
```
Comment (KaiSzuttor) [inline doc/sphinx/particles.rst]: the ```polymer``` in ```polymer_positions``` is redundant

Comment (KaiSzuttor) [reply doc/sphinx/particles.rst]: if you don't pollute the global namespace you can call the function with ```polymer.positions```

File: doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part3.ipynb [hunk 3]
```diff
@@ -94,8 +94,15 @@
     "\n",
     "# Setup polymer of part_id 0 with fene bond\n",
     "\n",
-    "polymer.create_polymer(N_P=1, MPC=mpc, bond=fene, bond_length=1,\n",
-    "                       start_pos=[16.0, 16.0, 16.0])\n",
+    "positions = polymer.polymer_positions(n_polymers=1,\n",
+    "                                      beads_per_chain=mpc,\n",
+    "                                      bond_length=1, seed=5642,\n",
+    "                                      min_distance=0.9)\n",
```
Comment (KaiSzuttor) [inline doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part3.ipynb]: why is the index necessary?

File: src/core/diamond.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+#include "utils.hpp"
+
+#include "utils/vec_rotate.hpp"
+using Utils::vec_rotate;
+
+/*************************************************************
+ * Functions                                                 *
+ * ---------                                                 *
+ *************************************************************/
+
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp:52]: Can you please come up with some self-explanatory function names?

File: src/core/polymer.hpp [hunk 5]
```diff
@@ -32,107 +32,43 @@
...
+Utils::Vector3d random_unit_vector(const std::function<double()> &generate_rn);
 
-/** C implementation of 'mindist \<part_id\> \<r_catch\>',<br>
- *  which returns the size of an array \<ids\> of indices of particles which are
- *  less than \<r_catch\> away from the position of the particle \<part_id\>.
+/** Returns the miminum distance between position pos and all existing
+ * particles.
  */
-int mindist3(PartCfg &, int part_id, double r_catch, int *ids);
+double mindist(PartCfg &partCfg, Utils::Vector3d pos);
```
Comment (KaiSzuttor) [inline src/core/polymer.hpp]: ```Utils::Vector3d const &pos```

File: src/core/polymer.hpp [hunk 6]
```diff
@@ -32,107 +32,43 @@
...
- *  collides
- *  with any other particle due to a minimum image distance smaller than
- *  \<shield\>.
- *  @param pos coordinates of particle to check
- *  @param shield minimum distance before it is defined as collision
- *  @param n_add number of additional coordinates to check
- *  @param add additional coordinates to check
- *  @return Returns '1' if there is a collision, '0' otherwise.
+/** Determines whether a given position pos is valid, i.e., it doesn't collide
+ * with existing or buffered particles, nor with existing constraints (if
```
Comment (KaiSzuttor) [inline src/core/polymer.hpp]: please add a description of the parameters

File: src/python/espressomd/polymer.pyx [hunk 7]
```diff
@@ -21,167 +21,112 @@ from __future__ import print_function, absolute_import
...
-    if _params["bond"]._bond_id == -1:
-        raise Exception(
-            "The bonded interaction passed as 'bond' keyword argument has not yet been added to the list of active bonds in Espresso.")
-    if _params["start_id"] < 0:
+    if not ((np.size(_params["start_positions"]) == 0)
+            or np.shape(_params["start_positions"]) == (_params["n_polymers"], 3)):
         raise ValueError(
-            "start_id has to be a positive Integer")
-    if not hasattr(_params["start_pos"], "__getitem__") or len(_params["start_pos"]) != 3:
+            "start_positions has to be an numpy array with shape (n_polymers, 3)")
```
Comment (KaiSzuttor) [inline src/python/espressomd/polymer.pyx]: an numpy

File: src/python/espressomd/polymer.pyx [hunk 8]
```diff
@@ -21,167 +21,112 @@ from __future__ import print_function, absolute_import
...
+        If set, this parameter defines the angle between adjacent bonds
+        withing a polymer.
+    start_positions : array_like :obj:`float`.
+        If set, this vector defines the start positions for the polymers, i.e.,
+        the position of each polymer's first monomer bead.
+        Here, a numpy array of shape (n_polymers, 3) is expected.
+    min_distance : :obj:`float`, optional
+        Minimum distance between all generated positions. Defaults to 0
+    respect_constraints : :obj:`bool`, optional
+        If True, the particle setup-up tries to obey previously defined constraints.
```
Comment (KaiSzuttor) [inline src/python/espressomd/polymer.pyx]: setup-up

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: doc/sphinx/particles.rst (hunk 2 above)
Comment (pkreissl) [reply doc/sphinx/particles.rst]: For me `positions` would be perfectly fine &ndash; still, some users might import the function via `from polymers import positions`, which might be problematic?!
An alternative could be something like `draw_postitions`, `get_postitions` (OK, that might imply that you get positions of already existing polymers), or `determine_positions` (which is probably too long). What is your opinion @RudolfWeeber?

File: doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part3.ipynb (hunk 3 above)
Comment (pkreissl) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part3.ipynb]: Which index do you mean?

File: doc/sphinx/particles.rst (hunk 2 above)
Comment (KaiSzuttor) [reply doc/sphinx/particles.rst]: You cannot hinder users from doing stupid things...
However, calling it ```polymer_positions``` does not describe what the function is doing

File: doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part3.ipynb (hunk 3 above)
Comment (KaiSzuttor) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part3.ipynb]: sorry wrong line, I meant the line below with ```positions[0]```. Why is the syntax different to the rest of the examples?

Comment (pkreissl) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part3.ipynb]: polymer_positions returns an array, which is basically a list of polymers, which themselves are a list of monomers, which in turn are a lists of xyz-coordinates:
```
[[[x01, y01, z01], [x02, y02, y03], ...]],
 [[x11, y11, z11], [x12, y12, y13], ...]],
 [[x21, y11, z21], [x22, y22, y23], ...]],
 ...                                     ]
```
This usually requires you to do two iterations:
```
positions = polymer.polymer_positions()
for polymer in positions:
    for monomer in polymer:
        ...
```
Here, I only have one polymer (```[ [ [x,y,x],[x,y,z],[x,y,z], ... [x,y,z] ] ]```), which I can directly get via index 0, so the outer for-loop becomes obsolete.

File: doc/sphinx/particles.rst (hunk 2 above)
Comment (pkreissl) [reply doc/sphinx/particles.rst]: OK, I'll go with you suggestion then.

File: src/core/polymer.hpp (hunk 5 above)
Comment (pkreissl) [reply src/core/polymer.hpp]: Makes clang:6.0 [CI test fail](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/104488):
```
/builds/espressomd/espresso/src/core/polymer.hpp:41:34: error: parameter 'pos' is const-qualified in the function declaration; const-qualification of parameters only has an effect in function definitions [readability-avoid-const-params-in-decls,-warnings-as-errors]
double mindist(PartCfg &partCfg, const Utils::Vector3d pos);
                                 ^~~~~~~~~~~
```

File: src/core/diamond.cpp [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+#include "constraints/ShapeBasedConstraint.hpp"
+#include "debug.hpp"
+#include "diamond.hpp"
+#include "global.hpp"
+#include "integrate.hpp"
+#include "random.hpp"
+#include "utils.hpp"
+
+#include "utils/vec_rotate.hpp"
+using Utils::vec_rotate;
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: ```vec_rotate``` is not used

File: src/core/diamond.cpp [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        return std::min(mindist, get_mi_vector(pos, p.r.p).norm2());
+      });
+
+  if (mindist < std::numeric_limits<double>::infinity())
+    return std::sqrt(mindist);
+  return -1.0;
+}
+
+double buf_mindist(double const pos[3], int n_add, double const *const add) {
+  double mindist = 30000.0, dx, dy, dz;
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: what is this magic number here?

File: src/core/diamond.cpp [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+      [&pos](double mindist, Particle const &p) {
+        return std::min(mindist, get_mi_vector(pos, p.r.p).norm2());
+      });
+
+  if (mindist < std::numeric_limits<double>::infinity())
+    return std::sqrt(mindist);
+  return -1.0;
+}
+
+double buf_mindist(double const pos[3], int n_add, double const *const add) {
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: please use ```Utils::Vector3d```. The name of the function is still not self-explanatory.

File: src/core/diamond.cpp [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+  return (1);
+}
+
+int create_counterions(PartCfg &partCfg, int N_CI, int part_id, int mode,
+                       double shield, int max_try, double val_CI, int type_CI) {
+  int n, cnt1, max_cnt;
+  double pos[3];
+
+  cnt1 = max_cnt = 0;
+  for (n = 0; n < N_CI; n++) {
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: ```for (int n```

File: src/core/diamond.cpp [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+    for (cnt1 = 0; cnt1 < max_try; cnt1++) {
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: ```for (int cnt1```

File: src/core/diamond.cpp [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+    dz -= std::round(dz / box_l[2]) * box_l[2];
+    mindist =
+        std::min(mindist, Utils::sqr(dx) + Utils::sqr(dy) + Utils::sqr(dz));
+  }
+  if (mindist < 30000.0)
+    return (sqrt(mindist));
+  return (-1.0);
+}
+
+int collides_with_particle(PartCfg &partCfg, double pos[3], double shield,
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: please return bool

File: src/core/diamond.cpp [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+      return (-1);
+    if (place_particle(part_id, pos) == ES_PART_ERROR)
+      return (-3);
+    set_particle_q(part_id, val_CI);
+    set_particle_type(part_id, type_CI);
+
+    part_id++;
+    max_cnt = std::max(cnt1, max_cnt);
+
+    POLY_TRACE(printf("C"); fflush(nullptr));
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: does this trace help?

File: src/core/diamond.cpp [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+  }
+  POLY_TRACE(printf(" %d->%d \n", cnt1, max_cnt));
+  if (cnt1 >= max_try)
+    return (-1);
+
+  return (std::max(max_cnt, cnt1));
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: why this extraneous brackets?

File: src/core/diamond.cpp [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+int create_diamond(PartCfg &partCfg, double a, double bond_length, int MPC,
+                   int N_CI, double val_nodes, double val_cM, double val_CI,
+                   int cM_dist, int nonet) {
+  int i, j, k, part_id, bond[2], type_bond = 0, type_node = 0, type_cM = 1,
+                                 type_nM = 1, type_CI = 2;
+  double pos[3], off = bond_length / sqrt(3);
+  double dnodes[8][3] = {{0, 0, 0}, {1, 1, 1}, {2, 2, 0}, {0, 2, 2},
+                         {2, 0, 2}, {3, 3, 1}, {1, 3, 3}, {3, 1, 3}};
+  int dchain[16][5] = {
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: please make this ```static constexpr```

File: src/core/diamond.cpp [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+
+  part_id = 0;
+  /* place 8 tetra-functional nodes */
+  for (i = 0; i < 8; i++) {
+    for (j = 0; j < 3; j++) {
+      dnodes[i][j] *= a / 4.;
+      pos[j] = dnodes[i][j];
+    }
+    if (place_particle(part_id, pos) == ES_PART_ERROR)
+      return (-3);
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: ```throw```?

File: src/core/diamond.cpp [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+  if (cnt1 >= max_try)
+    return (-1);
+
+  return (std::max(max_cnt, cnt1));
+}
+
+int create_diamond(PartCfg &partCfg, double a, double bond_length, int MPC,
+                   int N_CI, double val_nodes, double val_cM, double val_CI,
+                   int cM_dist, int nonet) {
+  int i, j, k, part_id, bond[2], type_bond = 0, type_node = 0, type_cM = 1,
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: please move the declarations right next to where the variables are used

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: src/core/diamond.cpp (hunk 9 above)
Comment (pkreissl) [reply src/core/diamond.cpp]: Point of this PR is not re-implementing the diamond code, which was part of the old polymer.cpp file. I just put it into a different file.

**Comment by pkreissl** [comment]
Point of this PR is not re-implementing the diamond code, which was part of the old polymer.cpp file. I just put it into a ~different~ separate file.

**Comment by KaiSzuttor** [comment]
No, but you can improve the code a lot with some trivial changes.

File: src/core/diamond.cpp (hunk 10 above)
Comment (pkreissl) [reply src/core/diamond.cpp]: Seems like a poor man's `std::numeric_limits<double>::infinity()`

File: src/core/polymer.hpp (hunk 5 above)
Comment (jngrad) [reply src/core/polymer.hpp]: the ampersand was missing, if you add it the error will go away

Comment (pkreissl) [reply src/core/polymer.hpp]: Added it just now. Thanks @jngrad!

File: src/core/diamond.cpp [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+  return std::max(max_cnt, cnt1);
+}
+
+int create_diamond(PartCfg &partCfg, double const a, double const bond_length,
+                   int const MPC, int const N_CI, double const val_nodes,
+                   double const val_cM, double const val_CI, int const cM_dist,
+                   int const nonet) {
+  Utils::Vector3d pos;
+  double const off = bond_length / sqrt(3);
+  double dnodes[8][3] = {
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: static constexpr

File: src/core/diamond.cpp [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+  for (int n = 0; n < N_CI; ++n) {
+    for (cnt1 = 0; cnt1 < max_try; ++cnt1) {
+      pos[0] = box_l[0] * d_random();
+      pos[1] = box_l[1] * d_random();
+      pos[2] = box_l[2] * d_random();
+      if ((mode != 0) or (mindist(partCfg, pos) > shield))
+        break;
+    }
+    if (cnt1 >= max_try)
+      throw std::runtime_error("Too many failed attpts finding valid position.");
```
Comment (jngrad) [inline src/core/diamond.cpp]: "attpts" -> "attempts"

File: src/core/diamond.cpp [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+      throw std::runtime_error("Failed to place particle.");
+    set_particle_q(part_id, val_CI);
+    set_particle_type(part_id, type_CI);
+
+    part_id++;
+    max_cnt = std::max(cnt1, max_cnt);
+  }
+  POLY_TRACE(printf(" %d->%d \n", cnt1, max_cnt));
+  if (cnt1 >= max_try)
+    throw std::runtime_error("Too many failed attpts finding valid position.");
```
Comment (jngrad) [inline src/core/diamond.cpp]: this branch will never be visited because the same condition at lines 68-69 throws

File: src/core/diamond.cpp [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+
+    part_id++;
+  }
+
+  /* place intermediate monomers on chains connecting the nodes */
+  for (int i = 0; i < 2 * 8; ++i) {
+    for (int k = 1; k <= MPC; ++k) {
+      for (int j = 0; j < 3; ++j)
+        pos[j] = dnodes[dchain[i][0]][j] + k * dchain[i][2 + j] * off;
+      if (place_particle(part_id, &pos[0]) == ES_PART_ERROR)
```
Comment (KaiSzuttor) [inline src/core/diamond.cpp]: a matter of taste but I find pos.data() more readable

File: src/core/diamond.cpp [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+    set_particle_type(part_id, type_CI);
+
+    part_id++;
+    max_cnt = std::max(cnt1, max_cnt);
+  }
+  POLY_TRACE(printf(" %d->%d \n", cnt1, max_cnt));
+  if (cnt1 >= max_try)
+    throw std::runtime_error("Too many failed attpts finding valid position.");
+
+  return std::max(max_cnt, cnt1);
```
Comment (jngrad) [inline src/core/diamond.cpp]: here simply `return max_cnt;`

**Review by jngrad — CHANGES_REQUESTED** [review]

File: src/core/diamond.cpp (hunk 20 above)
Comment (pkreissl) [reply src/core/diamond.cpp]: Can't do that, since dnodes is modified further down:
```
dnodes[i][j] *= a / 4.;
```

Comment (fweik) [reply src/core/diamond.cpp]:
```c++
auto const dnodes = a / 4. * Vector<Vector3d, 8>{{ ...
```

Comment (pkreissl) [reply src/core/diamond.cpp]: Did miss that `a` is just a constant factor... Thanks.

File: testsuite/python/polymer.py [hunk 25]
```diff
@@ -21,65 +21,185 @@
...
+        """
+        Assert that distance between monomers is bond_length.
+
+        """
+        for p in positions:
+            distances = np.linalg.norm(p[1:] - p[:-1], axis=1)
+            max_deviation = np.max(abs(distances - bond_length))
+            self.assertLess(max_deviation, 1e-10)
+
+    def assertBondAngle(self, positions, bond_angle, bond_length):
```
Comment (KaiSzuttor) [inline testsuite/python/polymer.py:54]: please adhere to the pep8 naming convention for methods: lowercase with underscores to separate words

Comment (pkreissl) [reply testsuite/python/polymer.py:54]: I tried to be consistent with the other asserts from pythons module [unittest](https://docs.python.org/2/library/unittest.html#assert-methods), which are all camel case (`assertEqual`, `assertRaisesRegex()`, etc.). I can change the nomencalture here, but personally I am in favour of consistency amongst the asserts.

**Comment by KaiSzuttor** [comment]
@pkreissl is this ready for review again?

**Comment by pkreissl** [comment]
Yes, it is. Thanks!

**Review by KaiSzuttor — APPROVED** [review]

**Comment by jngrad** [comment]
looking good, but do we really want to leave `diamond.cpp` at 0% coverage?

**Comment by KaiSzuttor** [comment]
I think this is a different issue that should be addressed in a separate issue/PR

**Review by jngrad — APPROVED** [review]

**Pull request merged**